File: 2025/day5/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int start;
    int end;
} Range;

int cmp(const void *a, const void *b) {
    Range *r1 = (Range*)a;
    Range *r2 = (Range*)b;
    return r1->start - r2->start;
}

int mergeRanges(Range ranges[], int n, Range merged[]) {

    qsort(ranges, n, sizeof(Range), cmp);

    int j = 0;
    merged[j] = ranges[0];

    for (int i = 1; i < n; i++) {

        if (ranges[i].start <= merged[j].end) {

            if (ranges[i].end > merged[j].end)
                merged[j].end = ranges[i].end;

        } else {
            j++;
            merged[j] = ranges[i];
        }
    }

    return j + 1;
}

int main() {

    FILE *f = fopen("inputs.txt", "r");

    Range ranges[100000];
    int numbers[100000];

    int rcount = 0, ncount = 0;
    int readingRanges = 1;

    char line[256];

    while (fgets(line, sizeof(line), f)) {

        if (line[0] == '\n') {
            readingRanges = 0;
            continue;
        }

        if (readingRanges) {
            int s, e;
            sscanf(line, "%d-%d", &s, &e);
            ranges[rcount++] = (Range){s,e};
        } else {
            int n;
            sscanf(line, "%d", &n);
            numbers[ncount++] = n;
        }
    }

    // PART 1
    int count = 0;

    for (int i = 0; i < ncount; i++) {
        for (int j = 0; j < rcount; j++) {
            if (numbers[i] >= ranges[j].start && numbers[i] <= ranges[j].end) {
                count++;
                break;
            }
        }
    }

    printf("Part1: %d\n", count);

    // PART 2
    Range merged[100000];

    int mcount = mergeRanges(ranges, rcount, merged);

    int total = 0;

    for (int i = 0; i < mcount; i++) {
        total += merged[i].end - merged[i].start + 1;
    }

    printf("Part2: %d\n", total);

    fclose(f);
}

File: 2025/day5/main.rs
use std::cmp::max;
use std::fs;

fn merge_intervals(mut ranges: Vec<(i64, i64)>) -> Vec<(i64, i64)> {
    ranges.sort_by_key(|r| r.0);

    let mut merged = Vec::new();

    for (start, end) in ranges {
        if merged.is_empty() {
            merged.push((start, end));
        } else {
            let last = merged.last_mut().unwrap();
            if start <= last.1 {
                last.1 = max(end, last.1);
            } else {
                merged.push((start, end));
            }
        }
    }
    merged
}

fn main() {
    let content = fs::read_to_string("inputs.txt").unwrap();

    // two mutable dynamic arrays
    let mut ranges = Vec::new();
    let mut numbers = Vec::new();

    // distinguish between ranges and numbers
    let mut reading_ranges = true;

    for line in content.lines() {
        if line.trim().is_empty() {
            reading_ranges = false;
            continue;
        }

        if reading_ranges {
            let parts: Vec<i64> = line.split('-').map(|x| x.parse().unwrap()).collect();

            ranges.push((parts[0], parts[1]));
        } else {
            numbers.push(line.parse::<i64>().unwrap());
        }
    }

    let mut count = 0;

    for n in numbers {
        for (start, end) in &ranges {
            if n >= *start && n <= *end {
                count += 1;
                break;
            }
        }
    }

    let merged = merge_intervals(ranges);

    let fresh_things: i64 = merged.iter().map(|(s, e)| e - s + 1).sum();

    println!("Part 1 = {}", count);
    println!("Part 2 = {}", fresh_things)
}
